### PR TITLE
feat!: web user interface

### PR DESCRIPTION
**Pull request by paularmstrong**
Work in progress. I'll update this as things continue to come together. 

Related: https://github.com/blakeblackshear/frigate/discussions/479

To run in development mode:

1. Build the docker images in the root of the repository `make amd64_all` (or appropriate for your system)
2. Create a config file in `config/config.yml`
3. Run the container: `docker run --rm --name frigate --privileged -v $PWD/config:/config:ro -v /etc/localtime:/etc/localtime:ro -p 5000:5000 frigate`
4. Run the dev ui: `cd web && npm run start` and open your browser to http://localhost:8080

For production mode, the `make` commands will build the HTML/JS/CSS. The built files (only) will be copied into the docker images and served from `/`

BREAKING CHANGES: This changes all HTTP endpoints from the python app to be served from `/api/` to avoid possible path collisions.

### Mask/Zone editor

- [x] Edit current masks & zones
- [x] Add a new mask when it does not previously exist
- [x] Add new zones
- [x] Remove mask/zone point

### CODE REVIEW

File: nginx/nginx.conf [hunk 1]
```diff
@@ -96,13 +96,19 @@ http {
             root /media/frigate;
         }
 
-        location / {
+        location /api/ {
+            add_header 'Access-Control-Allow-Origin' '*';
```
Comment (paularmstrong) [inline nginx/nginx.conf:100]: _note: temporary until we come up with a better way to allow the dev environment access to the api … unless we want to just open this up_

**Comment by tubalainen** [comment]
Fantastic! ! ❤️

**Comment by paularmstrong** [comment]
I think this is at a good starting point and "good enough" for a first pass in a beta release. @blakeblackshear can you give this a review and test it out?

**Comment by blakeblackshear** [comment]
Can you update this to point at the beta4 branch?

**Comment by paularmstrong** [comment]
> Can you update this to point at the beta4 branch?

Done!

**Comment by blakeblackshear** [comment]
Looks like some of the whitespace changes are conflicting now. Are you using VSCode? We should sync up editor configs.

Also, what version of node.js are you using? I can tweak this to do docker based builds.

**Comment by blakeblackshear** [comment]
Good enough of a starting point.

**Comment by paularmstrong** [comment]
> Looks like some of the whitespace changes are conflicting now. Are you using VSCode? We should sync up editor configs.

I'm using Sublime Text with `"trim_trailing_white_space_on_save": true`. Can't get myself to figure out vscode 😄 

> Also, what version of node.js are you using? I can tweak this to do docker based builds.

Either Node 12 or 14 should be okay (LTS releases are even numbers)

File: Makefile [hunk 2]
```diff
@@ -5,13 +5,16 @@ COMMIT_HASH := $(shell git log -1 --pretty=format:"%h")
 version:
 	echo "VERSION='0.8.0-$(COMMIT_HASH)'" > frigate/version.py
 
+web:
+	cd web && npm install && npm run build
```
Comment (pdecat) [inline Makefile:9]: Looks like a few deps are missing to build this:

```
npm run build

> frigate@ build /home/patrick/workspaces/homeassistant/frigate/frigate/web
> snowpack build

[snowpack] ! building source files...
[snowpack] ✘ file:///home/patrick/workspaces/homeassistant/frigate/frigate/web/src/index.css
  Error: Command failed with ENOENT: postcss
spawn postcss ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! frigate@ build: `snowpack build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the frigate@ build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/patrick/.npm/_logs/2021-01-13T16_48_21_842Z-debug.log
```

Running this resolves the issue:
```
npm install --save-dev @snowpack/plugin-postcss postcss postcss-cli
```

See https://www.npmjs.com/package/@snowpack/plugin-postcss

Comment (paularmstrong) [reply Makefile:9]: Oh no! I had thought they were direct dependencies, but looks like they're marked as peers. 

https://github.com/blakeblackshear/frigate/pull/518

**Comment by pdecat** [comment]
> BREAKING CHANGES: This changes all HTTP endpoints from the python app to be served from `/api/` to avoid possible path collisions.

Submitted https://github.com/blakeblackshear/frigate-hass-integration/pull/16 to take this change into account.

**Pull request merged**